Add verification for page frame allocator with safe permission-based memory access

ziqiaozhou opened the pull request:
This is a large PR since it starts to introduce the ghost tracked permission to reason unsafe memory access using unforgeable tracked memory permissions.

## Description of the PR

This PR add verification codes for allocator. 

### A new proof-related data type (verify_proof/src/frac_xxx.rs)

I introduced an unforgeable tracked memory permissions FracTypedPerm<T> based on verus's default memory permission [PointsTo<T>](https://github.com/verus-lang/verus/blob/main/source/vstd/raw_ptr.rs) (As for how the memory permission token is used, you can also refer to [VeriSMo](https://www.usenix.org/system/files/osdi24-zhou.pdf) to know more usage and properties it can provide, if you are really interested in). This permission type allows us to define the ownership of data when dealing with raw memory where Rust's existing ownership does not work well, especially for memory allocator. Only when the code has &mut FracTypedPerm<T>  and its shared == total_share, the code can modify the memory protected by the permission; When the code has &FracTypedPerm<T> where share > 0, the code can read the memory. This is a critical property for some memory area inside memory allocator.  All ghost functions provided by the type are fully verified to satisfying the properties we defined using state machine reasoning.

Other proof helper functions in verify_proof/src are not tighly related to the semantic of SVSM code and are just math proofs we used in svsm's proofs.

### Ghost data types defined for allocator verification.

`MemoryRegionPerms`: stores all memory permissions managed by MemoryRegion
`MRFreePerms`: stores all free memory permissions managed by MemoryRegion
`PageInfoDb`: Stores memory permissions pointing to PageInfo in reserved memory used by MemoryRegion. Those permissoins could be shared as readonly via FracTypedPerm when a page is allocated.
`MemRegionMapping`: A memory permission pointing to the global state of the memory region to ensure allocated memory and the global mem region shares the same mapping and address provenance.

### Verification of allocator (alloc_perm.verus.rs, alloc_free.verus.rs, alloc.rs)

With the above ghost permissions, in memory allocator, we need to prove the correct use of raw memory pointer and some invariants we defined for the memory region permissions stored in the memory allocator. verus checks the correct use of those memory permissions since we defined the trusted pointer TCB in tcb_ptr.rs to require a proper memory permission to read/write a raw ptr.

Another way to think about the verification here is that you can treat ghost permissions as a real memory block stored in the allocator, and then the allocator will return a memory block when allocating an instance from itself and will take a memory block when freeing an instance. When merging or splitting pages, the allocator will merge two blocks into one or split one into two. The use of those "memory block" must follow the rust ownership. With the ghost permission token, we can do this without introducing real memory overhead.

At a high level, it guarantees that

1. The MemoryRegion never returns a page for more than once.
2. The returned page must be a page managed by the region.
3. The page info stored is correct. If the page is allocated out of the allocator, the allocator should not be able to update the page info for that page until the page is freed. And thus, checking the page info is valid to know the actual type of the page.
4. A memory page to be freed must be an allocated memory. For any unsafe/unsafe code that tries to dealloc a memory page via API, that code must has ownership of the page.
5. The page counting is correct.
6. The code cannot change the definition of memory region (start, end, mapping) when some pages are allocated out of the allocator.

Among them, 1,2,4,6 are implicitly implied through the permission mechanism at any time point. 3 is partially implied by the permission and verified as an invariant in MRFreePerms::ens_perm_valid. 
Property 5 is verified for all pub or API functions by ensuring MemoryRegion::wf_perms which cannot be an invariant since MemoryRegion has some intermediate states (via some private functions) that can violate that property.

In addition, the verification helps to avoid some potential panic (e.g., Rust may panic if underflow/overflow happens) and provides proofs for some unreachable state. For all verified code, we can guarantee it will never cause overflow and underflow problems.  (If unverified code violates the preconditions we defined, it may still cause the problem, but we will incrementally ensure that will not happen).

### Encode and decode properties for some data types in allocator (alloc_type.verus.rs).

The allocator code also has many decode/encode function pairs, and I have proved the correct encode/decode relationship in SpecDecoderProof::proof_encode_decode (If a variable x can be encoded, the encoded value must be able to be decoded back to x) See that in alloc_type.verus.rs. 

## New information about the verification

In this PR, you will find three new syntax:
1. Defines extra **ghost** arguments passed to a function. The arguments does not exist in final compiled code.
```rust
#[verus_spec(with Tracked(x): Tracked<XType>)]
fn xyz() {}
```

2. Pass ghost arguments to a function call.
```rust
proof_with!(Tracked(x))
let _ = xyz();
```

3. Prove while declaring ghost/tracked variable that could be used in other proof blocks inside the same function
```rust
proof_decl!{
   let ghost x = 1usize;
   let tracked y = Map::tracked_empty();
}
```

4. Prove the loop with invariant
```rust
#[verus_spec(invariant true)] 
loop {}
```

5. Allow panic!() by enabling "vpanic" feature in builtin_macros and vstd.

In verus, the default setting disallows any panic to ensure the verified code is panic-free. But `svsm` has used panic in many places, and so we need to enable "vpanic" feature to allow panic to be used and also need to check the parameters passed to panic macro. 

## Changes

As before, verification-only files are
* under `verify_proof/src/` and `verify_externals/src/` 
* `kernel/src/*.verus.rs` are only used by verification.

Added executable files:
* [kernel/src/utils/safe_ptr.rs](https://github.com/coconut-svsm/svsm/compare/main...MSRSSP:svsm:verus-alloc-main#diff-854db1c5e0ea3471d8ebeffa0017dcf07e364bab029b32ef062cc43abfc87254) adds new raw pointer read/write function with a safe tracked permission.
Modified executable files:
* [kernel/src/address.rs](https://github.com/coconut-svsm/svsm/compare/main...MSRSSP:svsm:verus-alloc-main#diff-4345e2aa126d200ea1f34e2ef0bcffce704adc159e175aa570b28591752dbd11)
* [kernel/src/mm/address_space.rs](https://github.com/coconut-svsm/svsm/compare/main...MSRSSP:svsm:verus-alloc-main#diff-b4e8f094354e7e8101561be3695b473c69bf35e2c27f8780ed63e739831e64d3)
* [kernel/src/mm/alloc.rs](https://github.com/coconut-svsm/svsm/compare/main...MSRSSP:svsm:verus-alloc-main#diff-7eace8c7386d76dc9aa61d4f5d6e28af507602e921daad687f4064bf109d45c1)

NOTE: Inline comments will be left on executable code changes for further context.

## Verification Results
verification result could be found here at https://github.com/MSRSSP/svsm/actions/runs/14611944129/job/40991530406
Each commit is verifiable without issue (see per-commit CI result here at https://github.com/MSRSSP/svsm/pull/8)

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -802,39 +1462,73 @@ impl MemoryRegion {
+                proof_with!(Tracked(perm));
                 self.free_page_order(pfn, 0);
             }
-            PageInfo::Compound(ci) => {
-                let mask = (1usize << ci.order) - 1;
-                let start_pfn = pfn & !mask;
-                self.free_page_order(start_pfn, ci.order);
-            }
```
This can easily cause unexpected free and cause it hard to prove safety. Removed since it is not need. In current code, VMSA page is represented with two pages to guarantee it will free all pages together and does not need hack anymore.

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -802,39 +1462,73 @@ impl MemoryRegion {
+    )]
+    #[verus_verify(spinoff_prover)]
     fn free_page_order(&mut self, pfn: usize, order: usize) {
-        match self.try_to_merge_page(pfn, order) {
```
Fix clippy warning due to use of proof_with!() before try_to_merge_page

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -603,24 +1045,41 @@ impl MemoryRegion {
         });
-        self.write_page_info(pfn, pg);
-        Ok(self.start_virt + (pfn * PAGE_SIZE))
+        proof_with!(Tracked(perm));
+        self.allocate_pages_info(0, pg)
```
Reuse defined functions to avoid extra complexity.

ziqiaozhou commented on kernel/src/utils/safe_ptr.rs:
```diff
@@ -0,0 +1,124 @@
+        proof_decl! {
+            let tracked ptr_perm = perm.borrow();
+        }
+        vstd::raw_ptr::ptr_ref(self, verus_exec_expr!(Tracked(&ptr_perm)))
```
FracTypedPerm is built above PointsTo. vstd::raw_ptr::ptr_ref is a safe pointer wrapper from verus vstd lib (See https://github.com/verus-lang/verus/blob/main/source/vstd/raw_ptr.rs). It is just a wrapper to unsafe { &*ptr } that requires the correct use of tracked memory permission PointsTo<T>, to convert unsafe raw memory access to a verified safe behavior. Right now, our verification TCB is from verus-vstd lib, we can add `tcb` crate into the project in the future to list all trusted primitives with proper tracked variable as guardian (for unsafe functions(e.g., ptr_ref, pvalidate, rmpdajust, etc.).

ziqiaozhou commented on kernel/src/utils/safe_ptr.rs:
```diff
@@ -0,0 +1,124 @@
+            let tracked mut ptr_perm = perm.extract();
+            ptr_perm.leak_contents();
+        }
+        vstd::raw_ptr::ptr_mut_write(self, verus_exec_expr!(Tracked(&mut ptr_perm)), v);
```
Similar to ptr_ref, ptr_mut_write is a wrapper of core::ptr::write(ptr, v) with tracked memory permission to guard the usage.

ziqiaozhou commented:
@joergroedel , I updated the proof code to 
1. Reduce the proof block complexity with proof macros.
2. Support newer rust version. However, latest verus only supports 1.85.1 and older and so we need to manually modify the version in kernel/Cargo.toml to allow 1.85.1 before running `cargo verify`. I have tested it locally and in a forked repo with [CI](https://github.com/MSRSSP/svsm/actions/runs/15602192598/job/43944088563?pr=8). 
3. Avoid introducing extra dependencies from Verus in non-verification mode. My old change introduced some trusted executable code for pointer with provenance and tracked permission. Since the main branch has upgraded to 1.86, we can use rust's standard lib to track provenance, and I added two trusted functions in kernel/utils/tcb_ptr.rs with tracked memory permission checks to guarantee the safe read/write access to raw pointers.

Please let me know if you need any improvement or clarification to the change. Each commit is compliable and verifiable and so I can also dive this large PR to smaller ones if you prefer smaller PRs.

ziqiaozhou commented:
This PR has been rebased on the update verus https://github.com/coconut-svsm/svsm/pull/761.
All commits can pass the default CI and the verification CI in https://github.com/MSRSSP/svsm/pull/12.

stefano-garzarella commented:
> This PR has been rebased on the update verus #761. All commits can pass the default CI and the verification CI in [MSRSSP#12](https://github.com/MSRSSP/svsm/pull/12).

Okay, since there is a dependency on another, I'll move this to Draft, please mark it ready when we merge #761

ziqiaozhou commented:
> > This PR has been rebased on the update verus #761. All commits can pass the default CI and the verification CI in [MSRSSP#12](https://github.com/MSRSSP/svsm/pull/12).
> 
> Okay, since there is a dependency on another, I'll move this to Draft, please mark it ready when we merge #761

I had a conflict appointment and so cannot join the SVSM Development Call on July 23. But the PR is rebased and so should be ready for review. I did not push all commit one by one but they all passed the CI before rebasing and should all work well for standard build and verification.

ziqiaozhou commented on verify_proof/Cargo.toml:
```diff
@@ -8,14 +8,16 @@ edition = "2021"
 builtin = { workspace = true, optional = true }
 builtin_macros = { workspace = true }
-vstd = { workspace = true, optional = true }
+vstd = { workspace = true }
```
I can set it back to optional. I used to set vstd as required since this lib does not contains any executable code and should only be used for verification.

ziqiaozhou commented on verify_proof/Cargo.toml:
```diff
@@ -8,14 +8,16 @@ edition = "2021"
 seq-macro = "0.3"
 
 [lints]
-workspace = true
```
Re-enabled the lint again. I think I disabled it when developing due to some lint issue but now is resolved. Thanks for noticing that.

ziqiaozhou commented on kernel/src/types.verus.rs:
```diff
@@ -3,6 +3,7 @@
 // Copyright (c) Microsoft Corporation
 //
 // Author: Ziqiao Zhou <ziqiaozhou@microsoft.com>
+#[cfg(verus_keep_ghost)]
```
This is needed since the verus_stub does not do proper erasing for the content inside the verus macro. Currently the only places that we need to keep content in verus is for constant declaration.  If using verus builtin macro, the verus will only keep executable code and erase all ghost codes when cfg(not(verus_keep_ghost)).

ziqiaozhou commented:
> I tried to review, but commit message are very poor. Many of us know very little about verus, so please try to explain better in the commit messages what they do, at least to get an idea of what we are reviewing. Having a bulleted list (as I already pointed out in another PR), is not the standard practice we follow.
> 
> Also please check TODO, I saw at least 1 commit message with a TODO inside, not sure if it's on purpose.

I updated the comments and the PR description. It is a little bit long, but I hope it helps you to understand the PR. 

In addition, I also added some comments in code to help developers to understand some spec and struct I defined.

stefano-garzarella commented on kernel/src/address.verus.rs:
```diff
@@ -402,6 +403,10 @@ impl vstd::std_specs::ops::AddSpecImpl<InnerAddr> for PhysAddr {
 }
 
+} // verus!
+verus! {
```
why we need 2 `verus! } blocks here?

stefano-garzarella commented on verify_proof/src/bits.verus.rs:
```diff
@@ -388,4 +418,322 @@ pub broadcast proof fn lemma_bit_usize_shr_is_div(v: usize, n: usize)
+}
+
 } // verus!
+macro_rules! bit_xor_neighbor {
```
Why this is out of the `verus!` macro?

stefano-garzarella replied:
I guess because it's a macro. BTW can we move this on top of this file with others macro?

ziqiaozhou replied:
Yes. I move it above.

stefano-garzarella commented on verify_proof/src/frac_perm.rs:
```diff
@@ -0,0 +1,404 @@
+    }
+}
+
+} // verus!
+verus! {
```
Why closing and re-opening it?

ziqiaozhou replied:
Thanks for catching it.

stefano-garzarella commented on kernel/src/mm/address_space.rs:
```diff
@@ -26,12 +37,17 @@ impl FixedAddressMappingRange {
         } else {
             let size: usize = self.virt_end - self.virt_start;
-            if paddr >= self.phys_start + size {
+            if paddr - self.phys_start >= size {
```
IIUC this is a real fix on the code, so please put it at least in a separate commit, but even better in a different PR.

stefano-garzarella commented on Cargo.toml:
```diff
@@ -72,9 +72,9 @@ zerocopy = { version = "0.8.2", features = ["derive"] }
 # Verus repos
 builtin = { git = "https://github.com/verus-lang/verus", rev ="6c66898", default-features = false }
-builtin_macros = { git = "https://github.com/verus-lang/verus", rev ="6c66898", default-features = false }
+builtin_macros = { git = "https://github.com/verus-lang/verus", rev ="6c66898", features = ["vpanic"], default-features = false }
```
why we need this new feature?

I think these are things that we should describe in the commit message...

ziqiaozhou replied:
Documented it in comment at b48b27b9687457a49ab5245e5028c896939d5807. Verus expects verified code is panic free by default and so I need to enable panic feature to allow svsm to use panic. This is the first place we have panic inside verified code.

stefano-garzarella commented on kernel/src/address.rs:
```diff
@@ -358,14 +358,24 @@ impl VirtAddr {
+    )]
     pub fn as_ptr<T>(&self) -> *const T {
-        self.0 as *const T
+        core::ptr::with_exposed_provenance(self.0)
```
IMO, changes to the code should go in a different commits/PR with a good description in the commit message of the issue and maybe also on how it was found.

stefano-garzarella commented on kernel/src/address.rs:
```diff
@@ -358,14 +358,24 @@ impl VirtAddr {
+    )]
     pub fn as_mut_ptr<T>(&self) -> *mut T {
-        self.0 as *mut T
+        core::ptr::with_exposed_provenance_mut(self.0)
```
Ditto

stefano-garzarella reviewed (commented):
I did a quick review, let me summarize here what we should improve IMHO:
- avoid mixing `verus` changes and code changes in the same patch. If there is an issue, please fix it in a separate commit or even in another PR.
- break lines in commit description (e.g. https://github.com/coconut-svsm/svsm/pull/679/commits/424d39698c1a28f7692d7558bc82a4e12562fcff)
- explain changes in the commit description. I appreciated the PR description, but in the future, we will go through commits in our git history to debug issues or understand the verification code. So finding a commit like https://github.com/coconut-svsm/svsm/pull/679/commits/daeb5a9b1bdf86c3dd8f296eef49b1f2d540d07c that do ` 7 files changed, 412 insertions(+), 2 deletions(-)` fixing also a potential overflow in the code, with just 1 line description, is not great IMHO. How can we understand that in 1 or 2 years?
- please check the last commit description https://github.com/coconut-svsm/svsm/pull/679/commits/f9c79d84cad38b1bca6763369bbaa7e6067a0e7f
  - `Update verus version` where?
  - `Remove unncessary NEXT_MASK.` I looked in the diff, I can't see any `NEXT_MASK`
  - It looks like a summary of the entire series, so I'm confused... again, please put more effort in commit descriptions, describing the reasons for the change, and briefly what it changes.

ziqiaozhou commented:
> I did a quick review, let me summarize here what we should improve IMHO:
> 
> * avoid mixing `verus` changes and code changes in the same patch. If there is an issue, please fix it in a separate commit or even in another PR.
> * break lines in commit description (e.g. [424d396](https://github.com/coconut-svsm/svsm/commit/424d39698c1a28f7692d7558bc82a4e12562fcff))
> * explain changes in the commit description. I appreciated the PR description, but in the future, we will go through commits in our git history to debug issues or understand the verification code. So finding a commit like [daeb5a9](https://github.com/coconut-svsm/svsm/commit/daeb5a9b1bdf86c3dd8f296eef49b1f2d540d07c) that do ` 7 files changed, 412 insertions(+), 2 deletions(-)` fixing also a potential overflow in the code, with just 1 line description, is not great IMHO. How can we understand that in 1 or 2 years?
> * please check the last commit description [f9c79d8](https://github.com/coconut-svsm/svsm/commit/f9c79d84cad38b1bca6763369bbaa7e6067a0e7f)
>   
>   * `Update verus version` where?
>   * `Remove unncessary NEXT_MASK.` I looked in the diff, I can't see any `NEXT_MASK`
>   * It looks like a summary of the entire series, so I'm confused... again, please put more effort in commit descriptions, describing the reasons for the change, and briefly what it changes.

Thanks for the comments. I have splitted my changes in executable code from verus proofs. Now all comments start with $path: $description pattern and so it is clear `xxx.verus` path (e.g., kernel/mm/alloc.verus) indicate that change only add verus-related proofs/specs and does not really change exec code.  `verify_proof`/`verify_external` are all verus code without executable.

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -591,11 +589,6 @@ impl MemoryRegion {
     }
 
-    /// Calculates the end virtual address of the memory region.
-    fn end_virt(&self) -> VirtAddr {
```
Yes. The only place end_virt was used is in write_page_info (c6aea54e2290bb5df33bc5eed20fcfd31bec6e1c) and that place has been changed to avoid overflow.

ziqiaozhou replied:
Updated the commit comment.

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -557,11 +589,28 @@ impl MemoryRegion {
-        unsafe { self.start_virt.as_mut_ptr::<PageStorageType>().add(pfn) }
+        unsafe {
+            {
+                proof_with!(Tracked(self.perms.borrow().info_ptr_exposed));
```
This is to add a ghost variable to self.start_virt.as_mut_ptr::<PageStorageType>(). Currently, proof_with!(...) must be used together with as_mut_ptr so that verification tool will know the tracked ghost var is applied to that function call, and so I put it inside the same block with self.start_virt.as_mut_ptr::<PageStorageType>(). The original code has used `unsafe` and so I keep it there, but it does not need unsafe block in verified code since the safety is guarded by the exposed provenance tracked.

stefano-garzarella replied:
I see, thanks!

ziqiaozhou commented on kernel/src/mm/alloc.rs:
```diff
@@ -929,34 +1360,67 @@ impl MemoryRegion {
+        }
+
+        proof_with!(Tracked(&mut perm));
+        let merged = self.try_to_merge_page(pfn, order);
+        match merged {
```
This is a limitation of proof_with!() used inside verus_spec attribute macro. proof_with must be used close enough to try_to_merge_page where try_to_merge_page is used in a let-statement or is just a function call and could not be used in a complexed pattern like proof_with!(Tracked(&mut perm)) match elf.try_to_merge_page(pfn, order){ ... }. Such limitation might be resolved in a later version of verus but now this is how proof_with is supported in executable code.

stefano-garzarella commented on kernel/src/mm/alloc.rs:
```diff
@@ -431,11 +431,14 @@ impl MemoryRegion {
     /// The caller must provide a valid pfn, otherwise the returned pointer is
     /// undefined, as the compiler is allowed to optimize assuming there will
     /// be no arithmetic overflows.
-    unsafe fn page_info_mut_ptr(&mut self, pfn: usize) -> *mut PageStorageType {
```
@00xc @joergroedel since you touched this code, can you check if this is okay?

ziqiaozhou replied:
Based on the original comment, the function was marked unsafe since the returned pointer might be
undefined due to arithmetic overflows, and the compiler's optimization may assume there is no overflow and tried to use undefined ptr causing UB. That should be due to the original use of ptr::add(), Now I changed it to ptr::wrapping_add and thus should not cause undefined pointer. If start_virt or pfn is not valid, the result can still lead to invalid raw pointer which is true for VirtAddr::as_ptr(). Whether to keep both page_info_mut_ptr calculatin and ptr_read/ptr_write unsafe, or only do unsafe block for ptr read/write is trivial and does not matter too much in verus's world. It depends on the SVSM's preference. To me, I feel it is more natural to remove unsafe here since there is no unsafe functions called.

stefano-garzarella commented on kernel/src/mm/alloc.rs:
```diff
@@ -431,11 +431,14 @@ impl MemoryRegion {
-    unsafe fn page_info_mut_ptr(&mut self, pfn: usize) -> *mut PageStorageType {
+    #[expect(clippy::needless_pass_by_ref_mut)]
+    fn page_info_mut_ptr(&mut self, pfn: usize) -> *mut PageStorageType {
         // SAFETY: The start of the managed virtual address space always
```
This should be removed if we don't have the `unsafe` block anymore

stefano-garzarella commented on kernel/src/mm/alloc.rs:
```diff
@@ -445,11 +448,13 @@ impl MemoryRegion {
     /// be no arithmetic overflows.
-    unsafe fn page_info_ptr(&self, pfn: usize) -> *const PageStorageType {
+    fn page_info_ptr(&self, pfn: usize) -> *const PageStorageType {
         // SAFETY: The start of the managed virtual address space always
```
Ditto

stefano-garzarella commented on kernel/src/mm/alloc.rs:
```diff
@@ -600,15 +600,17 @@ impl MemoryRegion {
 
         // SAFETY: we have checked that the pfn is valid via check_pfn() above.
-        let info = unsafe { self.page_info_ptr(pfn).read() };
+        let info = unsafe { crate::utils::tcb_ptr::ptr_read(self.page_info_ptr(pfn)) };
```
As we usually do, can you add `use crate::utils::tcb_ptr::{ptr_read, ptr_write}` on top?

stefano-garzarella commented on kernel/src/mm/alloc.rs:
```diff
@@ -502,18 +503,28 @@ pub struct MemInfo {
 
 /// Memory region with its physical/virtual addresses, page count, as well
 /// as other details.
-#[derive(Debug, Default)]
```
Why `Default` is removed?

ziqiaozhou commented on kernel/src/address_inner.verus.rs:
```diff
@@ -18,7 +18,7 @@ pub spec const VADDR_MAX_BITS: nat = 48;
 pub spec const VADDR_LOWER_MASK: InnerAddr = 0x7fff_ffff_ffff as InnerAddr;
 
 #[verifier(inline)]
-pub spec const VADDR_UPPER_MASK: InnerAddr = !VADDR_LOWER_MASK;
```
Updated the comment. Also put a tiny proof there to indicate that it is still equivalent to !VADDR_LOWER_MASK.

stefano-garzarella commented:
@ziqiaozhou thanks! LGTM, except for the unneeded safety comment in https://github.com/coconut-svsm/svsm/pull/679#discussion_r2241958411
